File: 29Items-Projects/7Algorithms/3-Dynamic-Programming/java/src/main/java/com/example/dp/EditDistance.java
package com.example.dp;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class EditDistance {
    private EditDistance() {
    }

    public static Result memoized(String source, String target) {
        return new Result("Edit Distance", "memoization", solve(source, target, 0, 0, new HashMap<>()), Map.of());
    }

    private static int solve(String source, String target, int i, int j, Map<String, Integer> memo) {
        if (i == source.length()) {
            return target.length() - j;
        }
        if (j == target.length()) {
            return source.length() - i;
        }
        if (source.charAt(i) == target.charAt(j)) {
            return solve(source, target, i + 1, j + 1, memo);
        }
        String key = i + ":" + j;
        if (memo.containsKey(key)) {
            return memo.get(key);
        }
        int value = 1 + Math.min(
                solve(source, target, i + 1, j, memo),
                Math.min(solve(source, target, i, j + 1, memo), solve(source, target, i + 1, j + 1, memo)));
        memo.put(key, value);
        return value;
    }

    public static Result tabulated(String source, String target) {
        int[][] dp = buildTable(source, target);
        List<String> operations = new ArrayList<>();
        int i = source.length();
        int j = target.length();
        while (i > 0 || j > 0) {
            if (i > 0 && j > 0 && source.charAt(i - 1) == target.charAt(j - 1)) {
                i--;
                j--;
            } else if (i > 0 && dp[i][j] == dp[i - 1][j] + 1) {
                operations.add("delete " + source.charAt(i - 1));
                i--;
            } else if (j > 0 && dp[i][j] == dp[i][j - 1] + 1) {
                operations.add("insert " + target.charAt(j - 1));
                j--;
            } else {
                operations.add("replace " + source.charAt(i - 1) + "->" + target.charAt(j - 1));
                i--;
                j--;
            }
        }
        Collections.reverse(operations);
        return new Result("Edit Distance", "tabulation", dp[source.length()][target.length()], Map.of("operations", operations));
    }

    private static int[][] buildTable(String source, String target) {
        int[][] dp = new int[source.length() + 1][target.length() + 1];
        for (int i = 0; i <= source.length(); i++) {
            dp[i][0] = i;
        }
        for (int j = 0; j <= target.length(); j++) {
            dp[0][j] = j;
        }
        for (int i = 1; i <= source.length(); i++) {
            for (int j = 1; j <= target.length(); j++) {
                dp[i][j] = source.charAt(i - 1) == target.charAt(j - 1)
                        ? dp[i - 1][j - 1]
                        : 1 + Math.min(dp[i - 1][j], Math.min(dp[i][j - 1], dp[i - 1][j - 1]));
            }
        }
        return dp;
    }

    public static Result spaceOptimized(String source, String target) {
        int[] previous = new int[target.length() + 1];
        for (int j = 0; j <= target.length(); j++) {
            previous[j] = j;
        }
        for (int i = 1; i <= source.length(); i++) {
            int[] current = new int[target.length() + 1];
            current[0] = i;
            for (int j = 1; j <= target.length(); j++) {
                current[j] = source.charAt(i - 1) == target.charAt(j - 1)
                        ? previous[j - 1]
                        : 1 + Math.min(previous[j], Math.min(current[j - 1], previous[j - 1]));
            }
            previous = current;
        }
        return new Result("Edit Distance", "space optimized", previous[target.length()], Map.of());
    }
}

File: 29Items-Projects/7Algorithms/3-Dynamic-Programming/python/dp_algorithms/edit_distance.py
from __future__ import annotations

from functools import lru_cache

from dp_algorithms.common import DPResult


def memoized(source: str, target: str) -> DPResult:
    @lru_cache(maxsize=None)
    def solve(i: int, j: int) -> int:
        if i == len(source):
            return len(target) - j
        if j == len(target):
            return len(source) - i
        if source[i] == target[j]:
            return solve(i + 1, j + 1)
        return 1 + min(solve(i + 1, j), solve(i, j + 1), solve(i + 1, j + 1))

    return DPResult("Edit Distance", "memoization", solve(0, 0))


def tabulated(source: str, target: str) -> DPResult:
    rows, cols = len(source) + 1, len(target) + 1
    dp = [[0] * cols for _ in range(rows)]
    for i in range(rows):
        dp[i][0] = i
    for j in range(cols):
        dp[0][j] = j
    for i in range(1, rows):
        for j in range(1, cols):
            if source[i - 1] == target[j - 1]:
                dp[i][j] = dp[i - 1][j - 1]
            else:
                dp[i][j] = 1 + min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1])

    operations: list[str] = []
    i, j = len(source), len(target)
    while i > 0 or j > 0:
        if i > 0 and j > 0 and source[i - 1] == target[j - 1]:
            i -= 1
            j -= 1
        elif i > 0 and dp[i][j] == dp[i - 1][j] + 1:
            operations.append(f"delete {source[i - 1]}")
            i -= 1
        elif j > 0 and dp[i][j] == dp[i][j - 1] + 1:
            operations.append(f"insert {target[j - 1]}")
            j -= 1
        else:
            operations.append(f"replace {source[i - 1]}->{target[j - 1]}")
            i -= 1
            j -= 1
    operations.reverse()
    return DPResult("Edit Distance", "tabulation", dp[-1][-1], {"operations": operations})


def space_optimized(source: str, target: str) -> DPResult:
    previous = list(range(len(target) + 1))
    for i, left in enumerate(source, start=1):
        current = [i] + [0] * len(target)
        for j, right in enumerate(target, start=1):
            if left == right:
                current[j] = previous[j - 1]
            else:
                current[j] = 1 + min(previous[j], current[j - 1], previous[j - 1])
        previous = current
    return DPResult("Edit Distance", "space optimized", previous[-1])
